File: sahayog/sahayog/report/daily_sales_report/daily_sales_report.js
frappe.query_reports["Daily Sales Report"] = {
  filters: [
    {
      fieldname: "date",
      label: "Date",
      fieldtype: "Date",
      reqd: 1,
      default: frappe.datetime.get_today(),
    },
    {
      fieldname: "sol_id",
      label: "SOL ID",
      fieldtype: "Link",
      options: "Sahayog Branch",
      get_query: function () {
        if (
          frappe.user.has_role("Administrator") ||
          frappe.user.has_role("Sales Manager")
        ) {
          return {}; // no restrictions
        }
        return {
          filters: {
            sol_id: window.user_sol_id,
          },
        };
      },
    },
    {
      fieldname: "branch",
      label: "Branch",
      fieldtype: "Data",
      read_only: 1,
    },
  ],

  onload: async function (report) {
    const roles = frappe.user_roles;

    // Branch Manager default SOL/Branch
    if (roles.includes("Branch Manager")) {
      let user_branch = await frappe.call({
        method:
          "sahayog.scrm.doctype.dsr_remark.dsr_remark.get_user_branch_sol",
      });

      if (user_branch.message) {
        const { sol_id, branch } = user_branch.message;
        report.set_filter_value("sol_id", sol_id);
        report.set_filter_value("branch", branch);

        report.page.fields_dict.sol_id.df.read_only = 1;
        report.page.fields_dict.sol_id.refresh();
      }
    } else {
      report.page.fields_dict.sol_id.df.read_only = 0;
      report.page.fields_dict.sol_id.refresh();
    }

    // Admin & Sales Manager always get editable SOL ID
    if (
      frappe.user.has_role("Administrator") ||
      frappe.user.has_role("Sales Manager")
    ) {
      report.page.fields_dict.sol_id.df.read_only = 0;
      report.page.fields_dict.sol_id.refresh();
    }

    // ---------- FILTER UI ENHANCEMENT ----------

    // Add a light background + border around filter area
    const $form_area = report.page.wrapper.find(".page-form");
    $form_area.css({
      padding: "8px 10px",
      borderRadius: "6px",
      background: "#f9fafb",
      border: "1px solid #e5e7eb",
      marginBottom: "6px",
    });

    // Date field highlight on change
    const date_field = report.page.fields_dict.date;
    if (date_field && !date_field.__dsr_bound) {
      date_field.df.onchange = function () {
        let date_val = report.get_values().date;
        if (date_val) {
          frappe.show_alert(
            {
              message: __("Date changed to {0}", [date_val]),
              indicator: "blue",
            },
            3
          );
        }
        report.refresh();
      };
      date_field.__dsr_bound = true;
    }

    // SOL ID change handler - auto fetch branch + small alert
    report.page.fields_dict.sol_id.df.onchange = function () {
      let sol_id = report.get_values().sol_id;
      if (!sol_id) {
        report.set_filter_value("branch", "");
        return;
      }

      frappe.call({
        method: "frappe.client.get_value",
        args: {
          doctype: "Sahayog Branch",
          filters: { name: sol_id },
          fieldname: "branch",
        },
        callback: function (r) {
          if (r.message && r.message.branch) {
            report.set_filter_value("branch", r.message.branch);
            frappe.show_alert(
              {
                message: __("Branch set to {0} for SOL {1}", [
                  r.message.branch,
                  sol_id,
                ]),
                indicator: "green",
              },
              3
            );
          } else {
            report.set_filter_value("branch", "");
          }
        },
      });
    };

    // Remark button click binding
    $(document).off("click", ".remark-btn");
    $(document).on("click", ".remark-btn", function () {
      let emp_no = $(this).data("emp");
      let row = frappe.query_report.data.find(
        (r) => r.employee_number == emp_no
      );

      if (!row) {
        frappe.msgprint("Employee row not found.");
        return;
      }

      frappe.query_reports["Daily Sales Report"].open_employee_dialog(
        report,
        row
      );
    });

    // Hover style for remark button (add only once)
    if (!window.__dsr_hover_style__) {
      const style = document.createElement("style");
      style.innerHTML = `
        .remark-btn:hover {
          filter: brightness(1.08);
          transform: translateY(-1px);
        }
      `;
      document.head.appendChild(style);
      window.__dsr_hover_style__ = true;
    }

    // Export CSV button
    let export_btn = report.page.add_inner_button("Export CSV", () =>
      frappe.query_reports["Daily Sales Report"].export_csv(report)
    );

    export_btn
      .removeClass("btn-default btn-secondary")
      .addClass("btn-primary")
      .css({
        background: "rgb(22, 163, 74)",
        color: "#ffffff",
        border: "none",
      });
  },

  // Export CSV
  export_csv(report) {
    let data = frappe.query_report.data;

    if (!data || data.length === 0) {
      frappe.msgprint("No data available to export");
      return;
    }

    let csv_header = [
      "Employee Number",
      "Employee Name",
      "SOL ID",
      "Branch",
      "Designation",
      "Total Leads",
      "Converted Leads",
      "Followup Leads",
      "Not Interested Leads",
      "DSR Rating",
      "DSR Qualification",
      "Remark",
    ];

    let csv_rows = data.map((row) => [
      row.employee_number || "",
      row.employee_name || "",
      row.sol_id || "",
      row.branch || "",
      row.designation || "",
      row.total_leads || 0,
      row.converted_leads || 0,
      row.followup_leads || 0,
      row.not_interested_leads || 0,
      row.dsr_rating || "",
      row.dsr_qualification || "",
      row.existing_remark || "",
    ]);

    let csv_content =
      csv_header.join(",") + "\n" + csv_rows.map((e) => e.join(",")).join("\n");

    let blob = new Blob([csv_content], { type: "text/csv;charset=utf-8;" });
    let url = URL.createObjectURL(blob);

    let selected_date = report.get_values().date;
    let sol_id = report.get_values().sol_id || "all";
    let safe_date = selected_date.replace(/[^0-9\-]/g, "");
    let filename = `dsr_${safe_date}_${sol_id}.csv`;

    let link = document.createElement("a");
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);

    frappe.show_alert(`✔ CSV exported successfully as ${filename}`);
  },

  // Remark Dialog
  open_employee_dialog(report, row) {
    let dialog = new frappe.ui.Dialog({
      title: `Remark - ${row.employee_name} (${row.employee_number})`,
      fields: [
        {
          fieldname: "remark",
          label: "Remark",
          fieldtype: "Small Text",
          reqd: 1,
          default: row.existing_remark || "",
        },
      ],
      primary_action_label: "Save",
      primary_action(values) {
        frappe.call({
          method:
            "sahayog.scrm.doctype.dsr_remark.dsr_remark.create_or_update_dsr_remark",
          args: {
            date: report.get_values().date,
            employee: row.employee_number,
            employee_name: row.employee_name,
            sol_id: row.sol_id,
            branch: row.branch,
            designation: row.designation,
            total_leads: row.total_leads,
            converted_leads: row.converted_leads,
            followup_leads: row.followup_leads,
            not_interested_leads: row.not_interested_leads,
            dsr_rating: row.dsr_rating,
            dsr_qualification: row.dsr_qualification,
            remark: values.remark,
          },
          callback() {
            frappe.msgprint("✔ Remark Saved!");
            dialog.hide();
            report.refresh();
          },
        });
      },
    });

    dialog.show();
  },

  // Column formatter for badges
  formatter: function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);
    if (!data) return value;

    if (column.fieldname === "dsr_rating") {
      let color = data.dsr_rating_color || "gray";
      return `<span class="indicator ${color}">${frappe.utils.escape_html(
        data.dsr_rating || ""
      )}</span>`;
    }

    if (column.fieldname === "dsr_qualification") {
      let color = data.dsr_qualification_color || "gray";
      let cls = color === "green" ? "success" : "danger";
      return `<span class="badge badge-${cls}">${frappe.utils.escape_html(
        data.dsr_qualification || ""
      )}</span>`;
    }

    return value;
  },
};

File: sahayog/sahayog/report/daily_sales_report/daily_sales_report.py
import frappe
from frappe.utils import today


def execute(filters=None):
    if not filters:
        filters = {}

    selected_date = filters.get("date") or today()
    user = frappe.session.user
    user_roles = frappe.get_roles(user)

    # =====================================================
    # ROLE FILTERING
    # =====================================================
    if "Administrator" in user_roles or "Sales Manager" in user_roles:
        # Admin & Sales Manager see all employees
        employees = frappe.get_all(
            "Employee",
            fields=[
                "name",
                "employee_name",
                "employee_number",
                "designation",
                "sol_id",
                "user_id",
            ],
        )

    elif "Branch Manager" in user_roles:
        me = frappe.db.get_value(
            "Employee",
            {"user_id": user},
            ["name", "sol_id"],
            as_dict=True,
        )
        if not me:
            frappe.throw("No Employee assigned to current user.")

        # get branch
        branch = frappe.db.get_value(
            "Sahayog Branch", {"sol_id": me.sol_id}, "branch"
        )
        sol_ids = frappe.get_all(
            "Sahayog Branch", filters={"branch": branch}, pluck="sol_id"
        )

        employees = frappe.get_all(
            "Employee",
            filters={"sol_id": ["in", sol_ids]},
            fields=[
                "name",
                "employee_name",
                "employee_number",
                "designation",
                "sol_id",
                "user_id",
            ],
        )

    else:
        frappe.throw("You do not have permission to view this report.")

    # =====================================================
    # LOAD REMARKS MAP
    # =====================================================
    remarks_map = {}
    dsr_docname = frappe.db.get_value(
        "DSR Remark", {"date": selected_date}, "name"
    )

    if dsr_docname:
        dsr = frappe.get_doc("DSR Remark", dsr_docname)
        for row in dsr.dsr_employee_details:
            emp_no = frappe.db.get_value(
                "Employee", row.employee_id, "employee_number"
            )
            if emp_no:
                remarks_map[emp_no] = row.remark or ""

    # =====================================================
    # PRE-FETCH: branch names (bulk, keyed by sol_id)
    # =====================================================
    all_sol_ids = list({emp.sol_id for emp in employees if emp.sol_id})
    branch_rows = frappe.get_all(
        "Sahayog Branch",
        filters={"sol_id": ["in", all_sol_ids]},
        fields=["sol_id", "branch"],
    ) if all_sol_ids else []
    branch_map = {row.sol_id: row.branch for row in branch_rows}

    # PRE-FETCH: leads for all employees on selected_date (single query)
    all_user_ids = [emp.user_id for emp in employees if emp.user_id]
    if all_user_ids and all_sol_ids:
        lead_rows = frappe.db.sql(
            """
            SELECT lead_owner, sol_id, status
            FROM `tabLead`
            WHERE lead_owner IN ({user_placeholders})
              AND sol_id IN ({sol_placeholders})
              AND DATE(creation) = %s
            """.format(
                user_placeholders=",".join(["%s"] * len(all_user_ids)),
                sol_placeholders=",".join(["%s"] * len(all_sol_ids)),
            ),
            tuple(all_user_ids) + tuple(all_sol_ids) + (selected_date,),
            as_dict=True,
        )
    else:
        lead_rows = []

    # Group leads by (lead_owner, sol_id)
    leads_map = {}
    for row in lead_rows:
        key = (row.lead_owner, row.sol_id)
        leads_map.setdefault(key, []).append(row.status)

    # =====================================================
    # DATA BUILDING
    # =====================================================
    data = []

    for emp in employees:
        branch_name = branch_map.get(emp.sol_id) or "Not Mapped"

        # Accurate per-day leads (lookup from pre-fetched map)
        statuses = leads_map.get((emp.user_id, emp.sol_id), [])

        total = len(statuses)
        converted = statuses.count("Converted")
        followup = statuses.count("Follow Up")
        notint = statuses.count("Not Interested")

        # Rating + color
        if converted >= 1:
            rating = "Good"
            rating_color = "green"
        elif followup >= 4 and converted == 0:
            rating = "Average"
            rating_color = "orange"
        else:
            rating = "Bad"
            rating_color = "red"

        qualify = "Qualified" if total >= 10 else "Disqualified"
        qualify_color = "green" if qualify == "Qualified" else "red"

        existing_remark = remarks_map.get(emp.employee_number, "")

        # =========================================
        # BUTTON RENDERING
        # =========================================
        if not existing_remark:
            btn_label = "Add Remark"
            remark_cell = (
                f"<div style='text-align:center; display:flex; align-items:center; justify-content:center;'>"
                f"<button class='remark-btn' "
                f"data-emp='{emp.employee_number}' "
                f"style='background:rgb(59 130 246); color:#ffffff; border:none; "
                f"border-radius:6px; padding:4px 10px; font-size:12px; cursor:pointer; "
                f"transition:all 0.15s ease;'>"
                f"{btn_label}</button>"
                f"</div>"
            )
        else:
            btn_label = "View Remark"
            remark_cell = (
                f"<div style='text-align:center; display:flex; align-items:center; justify-content:center;'>"
                f"<button class='remark-btn' "
                f"data-emp='{emp.employee_number}' "
                f"style='background:rgb(22 163 74); color:#ffffff; border:none; "
                f"border-radius:6px; padding:4px 10px; font-size:12px; cursor:pointer; "
                f"transition:all 0.15s ease;'>"
                f"{btn_label}</button>"
                f"</div>"
            )

        data.append(
            {
                "sol_id": emp.sol_id,
                "branch": branch_name,
                "employee_number": emp.employee_number,
                "employee_name": emp.employee_name,
                "designation": emp.designation,
                "total_leads": total,
                "converted_leads": converted,
                "followup_leads": followup,
                "not_interested_leads": notint,
                "dsr_rating": rating,
                "dsr_rating_color": rating_color,
                "dsr_qualification": qualify,
                "dsr_qualification_color": qualify_color,
                "existing_remark": existing_remark,
                "remarks": remark_cell,
            }
        )

    # SORTING
    data.sort(key=lambda x: x["total_leads"], reverse=True)

    # =====================================================
    # COLUMNS
    # =====================================================
    columns = [
        {"label": "SOL ID", "fieldname": "sol_id", "fieldtype": "Data", "width": 90},
        {
            "label": "Remarks",
            "fieldname": "remarks",
            "fieldtype": "HTML",
            "width": 160,
            "height": 100,
        },
        {"label": "Branch", "fieldname": "branch", "fieldtype": "Data", "width": 120},
        {
            "label": "Emp ID",
            "fieldname": "employee_number",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": "Employee Name",
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 150,
        },
        {
            "label": "Designation",
            "fieldname": "designation",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "Total Leads",
            "fieldname": "total_leads",
            "fieldtype": "Int",
            "width": 90,
        },
        {
            "label": "Converted",
            "fieldname": "converted_leads",
            "fieldtype": "Int",
            "width": 90,
        },
        {
            "label": "Follow-up",
            "fieldname": "followup_leads",
            "fieldtype": "Int",
            "width": 90,
        },
        {
            "label": "Not Interested",
            "fieldname": "not_interested_leads",
            "fieldtype": "Int",
            "width": 130,
        },
        {
          "label": "DSR Rating",
          "fieldname": "dsr_rating",
          "fieldtype": "Data",
          "width": 90,
        },
        {
          "label": "DSR Qualification",
          "fieldname": "dsr_qualification",
          "fieldtype": "Data",
          "width": 130,
        },
        # hidden helper fields for color
        {
          "label": "Rating Color",
          "fieldname": "dsr_rating_color",
          "fieldtype": "Data",
          "hidden": 1,
        },
        {
          "label": "Qual Color",
          "fieldname": "dsr_qualification_color",
          "fieldtype": "Data",
          "hidden": 1,
        },
    ]

    return columns, data
